packages/backend-tasks/src/tasks/TaskScheduler.ts

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { DatabaseManager, getRootLogger } from '@backstage/backend-common';
+import { Config } from '@backstage/config';
+import { memoize } from 'lodash';
+import { Duration } from 'luxon';
+import { Logger } from 'winston';
+import { migrateBackendTasks } from '../database/migrateBackendTasks';
+import { PluginTaskSchedulerImpl } from './PluginTaskSchedulerImpl';
+import { PluginTaskSchedulerJanitor } from './PluginTaskSchedulerJanitor';
+import { PluginTaskScheduler } from './types';
+
+/**
+ * Deals with the scheduling of distributed tasks.
+ *
+ * @public
+ */
+export class TaskScheduler {
+  static fromConfig(
+    config: Config,
+    options?: {
+      databaseManager?: DatabaseManager;
+      logger?: Logger;
+    },
+  ): TaskScheduler {
+    const databaseManager =
+      options?.databaseManager ?? DatabaseManager.fromConfig(config);
+    const logger = (options?.logger || getRootLogger()).child({
+      type: 'taskManager',
+    });
+    return new TaskScheduler(databaseManager, logger);
+  }
+
+  constructor(
+    private readonly databaseManager: DatabaseManager,
+    private readonly logger: Logger,
+  ) {}
+
+  /**
+   * Instantiates a task manager instance for the given plugin.
+   *
+   * @param pluginId - The unique ID of the plugin, for example "catalog"
+   * @returns A {@link PluginTaskScheduler} instance
+   */
+  forPlugin(pluginId: string): PluginTaskScheduler {
+    const databaseFactory = memoize(async () => {
```

Comment:
Should this `memoize` have the cache key resolver argument passed in as well? AFAIK the default implementation uses the first arg passed to the memoized function as the cache key which in this case would always be `undefined`.

Reply:
The intent here is to make sure that the inner function is only called once, no matter how many times the result is called. Giving no cache key argument should have that effect as far as I understand it. But I notice now that lodash also has a `once` decorator, which more clearly has the desired semantic, so I'll switch to using that.